### ISSUE TITLE
Feature: Focus option for top level tests

### ISSUE THREAD

**Issue by sksamuel** [kotlintest/kotlintest#329]

**Comment by sksamuel** [comment]
It's in 3.1.1 if you want to use it :)

**Comment by sksamuel** [comment]
Should work for all, but only top level tests. @dave08

**Comment by dave08** [comment]
It would be great if it could work for groupings and top level tests, since in KotlinTest there's no IDE arrows like in Spek and JUnit besides each test, so one has to run the whole class or disable all the other tests just to concentrate on the one test he/she's developing... @sksamuel

**Comment by sksamuel** [comment]
If you put f: in front of your test name then only that test (and nested tests under it) will be executed @dave08 
Eg,

See

https://github.com/kotlintest/kotlintest/blob/master/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/FocusTest.kt

I'm not sure what you mean by your second comment about ! and options.

**Comment by dave08** [comment]
Say:
```kotlin
class FooTest : DescribeSpec({
	describe("A foo") {
		context("!Planned feature (not yet working on this)") {
			// ... All the tests are skipped/disabled under here because of the ! prefix
		}

		context("Feature already finished") {
			// ... These tests should run when there are no `f:`s
		}

		context("f:Current feature") {
			// ... Only these tests should run as long as there's `f:` here
		}
	}
})
```
@sksamuel

**Comment by sksamuel** [comment]
I see, no you can't do that. The reason is, inner tests are not known until runtime. So it would break the way the test runners work.

You can put the f: on the describe block.

**Comment by dave08** [comment]
So how does providing the `options()` work in `it()`, one could disable the test there... if when adding the test you check the label string to make the same decision it would be possible no? At least for `!`... I don't think I saw a parameter for `f:` in the `options()`, maybe that's what you mean is impossible? I was actually wondering why there's no `focus` param there...

**Comment by sksamuel** [comment]
In order to determine which test to run, it has to check all of them in the same scope.

If you do:

```kotlin
  test("test 1") { }
  test("f: test 2") { }
```

Then test 1 needs to be skipped because you've focused test 2. So it needs to scan ahead. We can't do this in nested scopes as the test closures are executed lazily, so that variable initialisation etc, works properly.

**Comment by dave08** [comment]
So maybe have a `f2:` on each level scope to focus on the second item on the level under it? A bit harder to use, but may be easier to know at run time what to run @sksamuel.

**Comment by sksamuel** [comment]
That's not the problem.
The problem is that you have to look ahead. If you mark test 20 with f: or f2, then the engine has to skip the first 19. It can only know to do this if it has already scanned the complete list of test names.

**Comment by dave08** [comment]
I mean f2 would be on the parent to indicate the second test or group under it is focused

**Comment by sksamuel** [comment]
You mean f2 would indicate the 2nd test was focused, and f3 the third test and so on?

**Comment by dave08** [comment]
Yes, the second test under the indicated scope, but a clearer naming might be chosen to indicate not focusing on the item with the f but its child maybe `f-2:` (like the freespec group indicator `-` ) or `2>` or `>2` ?

It might get complicated in parameterized tests, I don't know if it would be possible to run the whole parameter set (not just the 4th one for eg.), which makes I think more sense in the devs workflow.

**Comment by TAGC** [comment]
> I see, no you can't do that. The reason is, inner tests are not known until runtime. So it would break the way the test runners work.

Yeah, this is pretty annoying. Using DescribeSpec, I need to literally comment out every test but the one I want to focus on in order to isolate it.

**Comment by LeoColman** [comment]
I believe that with the intellij plugin, this will be somewhat easier,
right?

**Comment by sksamuel** [comment]
Yes, you can run in the inner test with the plugin.
It will actually make focus a bit redundant.

**Comment by TAGC** [comment]
Oh, I didn't realise there was a plugin, that's neat.

I just installed it and now notice there's arrows in the IDE beside each individual test case (within a `DescribeScope`). However, clicking any of them just runs all the test cases, the same as if I click the arrow beside the top-level "describe(...)". Should the plugin support running individual nested test cases?

This is what I mean:

![idea64_2019-02-20_12-11-39](https://user-images.githubusercontent.com/1588951/53090962-b5f12100-3508-11e9-8d08-e7376ecea57a.png)

**Comment by LeoColman** [comment]
> It will actually make focus a bit redundant.

Not redundant if you consider people still use eclipse

**Comment by sksamuel** [comment]
Do they? In 2019 :)

Yes fair point.

**Comment by sksamuel** [comment]
@TAGC You're the second person to say this. Although the "run" tool tip is wrong (I don't know how to fix that currently), for me it does only run the correct single test. Can you try it, and take a screenshot of the run configuration page.

**Comment by TAGC** [comment]
Sure, the run configuration (this is used regardless of which green arrow I click):

![idea64_2019-02-20_12-39-08](https://user-images.githubusercontent.com/1588951/53092489-d58a4880-350c-11e9-8db5-4a2663eea7f1.png)

And here's the test output pane:

![idea64_2019-02-20_12-40-40](https://user-images.githubusercontent.com/1588951/53092498-d7540c00-350c-11e9-9d5c-a75756df262f.png)

Here's the relevant logging output:

```
40 [main] DEBUG KotlinTestEngine - configurationParameters=LauncherConfigurationParameters []
40 [main] DEBUG KotlinTestEngine - uniqueId=[engine:kotlintest]
60 [main] DEBUG KotlinTestEngine - LauncherDiscoveryRequest [
classpathRootSelectors=[]
classpathResourceSelectors=[]
classSelectors=[com.mycompany.testmaster.domain.nodes.CompositeNodeTests]
methodSelectors=[]
directorySelectors=[]
fileSelectors=[]
moduleSelectors=[]
packageSelectors=[]
uniqueIdSelectors=[]
uriSelectors=[]
engineFilters=[]
postDiscoveryFilters=[]
classnameFilters=[]
packageNameFilters=[]
]
120 [main] DEBUG TestDiscovery - Loaded 1 classes from classnames...
395 [main] DEBUG TestDiscovery - ...which has filtered to 1 non abstract classes
395 [main] DEBUG TestDiscovery - 1 classes after applying discovery extensions405 [main] DEBUG KotlinTestEngine - configurationParameters=LauncherConfigurationParameters []
405 [main] DEBUG KotlinTestEngine - uniqueId=[engine:kotlintest]
405 [main] DEBUG KotlinTestEngine - LauncherDiscoveryRequest [
classpathRootSelectors=[]
classpathResourceSelectors=[]
classSelectors=[com.mycompany.testmaster.domain.nodes.CompositeNodeTests]
methodSelectors=[]
directorySelectors=[]
fileSelectors=[]
moduleSelectors=[]
packageSelectors=[]
uniqueIdSelectors=[]
uriSelectors=[]
engineFilters=[]
postDiscoveryFilters=[]
classnameFilters=[]
packageNameFilters=[]
]
415 [main] DEBUG KotlinTestEngine - JUnit execution request [configurationParameters=LauncherConfigurationParameters []; rootTestDescriptor=KotlinTestEngineDescriptor: [engine:kotlintest]]
415 [main] DEBUG JUnitTestRunnerListener - Engine started; classes=[[class com.mycompany.testmaster.domain.nodes.CompositeNodeTests]]
415 [main] DEBUG TestEngine - Submitting 1 specs
415 [main] DEBUG TestEngine - Waiting for spec execution service to terminate
437 [kotlintest-engine-0] DEBUG JUnitTestRunnerListener - prepareSpec [Description(parents=[], name=CompositeNodeTests)]447 [kotlintest-test-executor-0] DEBUG JUnitTestRunnerListener - Creating test case descriptor Description(parents=[CompositeNodeTests], name=Describe: Composite node)/Container
457 [kotlintest-test-executor-0] DEBUG JUnitTestRunnerListener - Notifying junit of start event [engine:kotlintest]/[spec:CompositeNodeTests]/[test:Describe%3A Composite node]457 [kotlintest-test-executor-0] DEBUG JUnitTestRunnerListener - Creating test case descriptor Description(parents=[CompositeNodeTests, Describe: Composite node], name=Scenario: immediately executes once and terminates if it has no inputs and no outputs)/Test
457 [kotlintest-test-executor-0] DEBUG JUnitTestRunnerListener - Notifying junit of start event [engine:kotlintest]/[spec:CompositeNodeTests]/[test:Describe%3A Composite node]/[test:Scenario%3A immediately executes once and terminates if it has no inputs and no outputs]
634 [kotlintest-engine-0] DEBUG JUnitTestRunnerListener - Notifying junit of test case completion [engine:kotlintest]/[spec:CompositeNodeTests]/[test:Describe%3A Composite node]=TestResult(status=Failure, error=java.lang.AssertionError: Expected exception java.lang.IllegalStateException but no exception was thrown, reason=null, metaData={})
```

**Comment by sksamuel** [comment]
Thanks.
It's not actually created it as a kotlintest run but a standard junit 5.
Is there an option to create a KT test manually if you click the plus icon ?
